File: wintouch/FbItemMessage.cpp
#include "FbItemMessage.h"

#include <msclr\marshal_cppstd.h>

#include "MainWindow.h"

FbItemMessage::FbItemMessage(const TheUI::FbEventData& msg)
	: FbMessage(TableName(), msg)
	, name()
	, categoryId()
	, description()
	, price(-1.0)
	, qtyPerBox(-1)
	, units()
	, supplierId()
{ }

int FbItemMessage::Respond()
{
	int ret = -1;

	if (!isParsed && Parse() != 0)
	{
		ParseFailedMessage();
	}
	else if (OperationType() == TheUI::Fb_Operations::fb_add || OperationType() == TheUI::Fb_Operations::fb_edit)
	{
		System::String^ idCli         = msclr::interop::marshal_as<System::String^>(id);
		System::String^ nameCli       = msclr::interop::marshal_as<System::String^>(name);
		System::String^ catCli        = msclr::interop::marshal_as<System::String^>(categoryId);
		System::String^ descrCli      = msclr::interop::marshal_as<System::String^>(description);
		System::String^ unitsCli      = msclr::interop::marshal_as<System::String^>(units);
		System::String^ supplierIdCli = msclr::interop::marshal_as<System::String^>(supplierId);
		TheUI::ItemPropEntryFb item(nameCli, catCli, descrCli, price, qtyPerBox, unitsCli, supplierIdCli);
		ManagedCode::ManagedGlobals::w->AddProperties(idCli, %item);
		ret = 0;
	}
	else if (OperationType() == TheUI::Fb_Operations::fb_delete)
	{
		System::String^ idCli = msclr::interop::marshal_as<System::String^>(id);
		ManagedCode::ManagedGlobals::w->RemoveItemProperties(idCli);
		ret = 0;
	}
	else UnsupportedOperationMessage();

	return ret;
}

int FbItemMessage::Parse()
{
	if (OperationType() == TheUI::Fb_Operations::fb_add || OperationType() == TheUI::Fb_Operations::fb_edit)
	{
		if (!isParsed && ParseId(NAME)		  == 0) { name		  =      Data()         ; }
		if (!isParsed && ParseId(CATEGORY_ID) == 0) { categoryId  =      Data()         ; }
		if (!isParsed && ParseId(DESCRIPTION) == 0) { description =		 Data()			; }
		if (!isParsed && ParseId(PRICE)		  == 0) { price		  = atof(Data().substr(1).c_str()); }
		if (!isParsed && ParseId(QTY_PER_BOX) == 0) { qtyPerBox	  = atoi(Data().substr(1).c_str()); }
		if (!isParsed && ParseId(UNITS)		  == 0) { units		  =      Data()         ; }
		if (!isParsed && ParseId(SUPPLIER_ID) == 0) { supplierId  =      Data()         ; }
	}
	else if (OperationType() == TheUI::Fb_Operations::fb_delete)
	{
		ParseId();
	}
	return isParsed ? 0 : -1;
}

const std::string FbItemMessage::NAME		 = "Name";
const std::string FbItemMessage::CATEGORY_ID = "CategoryId";
const std::string FbItemMessage::DESCRIPTION = "Description";
const std::string FbItemMessage::PRICE		 = "Price";
const std::string FbItemMessage::QTY_PER_BOX = "QtyPerBox";
const std::string FbItemMessage::UNITS		 = "Units";
const std::string FbItemMessage::SUPPLIER_ID = "SupplierId";

File: TheUI/FbMessages/FbItemMessage.cs
using System;

namespace TheUI
{
    class FbItemMessage : FbMessage, IItemPropEntryFb
    {
        public FbItemMessage(FbEventData msg) : base(msg)
        {
            Name        = "";
            CategoryId  = "";
            Description = "";
            Units       = "";
            SupplierId  = "";
            Price       = -1.0;
            QtyPerBox   = -1;
            Image       = "";
            Parse();
        }

        public const string TABLE_NAME = "items";

        protected override sealed int Parse()
        {
            if (OPERATION == Fb_Operations.fb_add || OPERATION == Fb_Operations.fb_edit)
            {
                if (!IsParsed && ParseId(NAME)        == 0) { Name        = DATA; }
                if (!IsParsed && ParseId(CATEGORY_ID) == 0) { CategoryId  = DATA; }
                if (!IsParsed && ParseId(DESCRIPTION) == 0) { Description = DATA; }
                if (!IsParsed && ParseId(PRICE)       == 0)
                {
                    double price = -1.0;
                    if (Double.TryParse(DATA.Substring(1), out price)) Price = price;
                }
                if (!IsParsed && ParseId(QTY_PER_BOX) == 0)
                {
                    int qty = -1;
                    if (Int32.TryParse(DATA.Substring(1), out qty)) QtyPerBox = qty;
                }
                if (!IsParsed && ParseId(UNITS)       == 0) { Units      = DATA; }
                if (!IsParsed && ParseId(SUPPLIER_ID) == 0) { SupplierId = DATA; }
                if (!IsParsed && ParseId(IMAGE)       == 0) { Image      = DATA; }
            }
            else if (OPERATION == Fb_Operations.fb_delete)
            {
                ParseId();
            }
            return IsParsed ? 0 : -1;
        }

        public string Name        { get; set; }
        public string CategoryId  { get; set; }
        public string Description { get; set; }
        public double Price       { get; set; }
        public int    QtyPerBox   { get; set; }
        public string Units       { get; set; }
        public string SupplierId  { get; set; }
        public string Image       { get; set; }

        private const string NAME        = "Name";
        private const string CATEGORY_ID = "CategoryId";
        private const string DESCRIPTION = "Description";
        private const string PRICE       = "Price";
        private const string QTY_PER_BOX = "QtyPerBox";
        private const string UNITS       = "Units";
        private const string SUPPLIER_ID = "SupplierId";
        private const string IMAGE       = "Image";
    }
}
